python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,1111 @@
+                if len(vms) > 0:
+                    cli_logger.info("Frozen VM already exists. Not creating again.")
+
+                    vm_id = vms[0].vm
+                    status = self.vsphere_automation_sdk_client.vcenter.vm.Power.get(
+                        vm_id
+                    )
+
+                    if status.state != HardPower.State.POWERED_ON:
+                        cli_logger.info("Frozen VM is off. Powering it ON")
+                        self.vsphere_automation_sdk_client.vcenter.vm.Power.start(vm_id)
+                        cli_logger.info("vm.Power.start({})".format(vm_id))
+                    return
+
+                cli_logger.info("Creating frozen VM")
+                self._return = self.outer_obj.create_ovf_node(self.node_config, vm_name)
+            except Exception as e:
+                cli_logger.info("Failed creating frozen VM {}".format(self.target_name))
+                self.thread_exception = e
+
+        def join(self):
+            return super().join()
+
+    @staticmethod
+    def bootstrap_config(cluster_config):
+        return bootstrap_vsphere(cluster_config)
+
+    def non_terminated_nodes(self, tag_filters):
+        """For vSphere, non_terminated nodes are:
+        1. Nodes in creating state.
+        2. Nodes already created and powered on.
+        """
+        with self.lock:
+            nodes = []
+            cli_logger.info("Getting non terminated nodes...")
+            vms = self.vsphere_automation_sdk_client.vcenter.VM.list()
+            cli_logger.info(f"Got {len(vms)} non terminated nodes.")
+            filters = tag_filters.copy()
+            if TAG_RAY_CLUSTER_NAME not in tag_filters:
+                filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+
+            for vm in vms:
+                vm_id = vm.vm
+                yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm.vm)
+
+                matched_tags = {}
+                all_tags = {}
+                # If the VM has a tag from tag_filter then select it
+                for (
+                    tag_id
+                ) in self.vsphere_automation_sdk_client.tagging.TagAssociation.list_attached_tags(
+                    yn_id
+                ):
+                    vsphere_vm_tag = self.vsphere_automation_sdk_client.tagging.Tag.get(
+                        tag_id=tag_id
+                    ).name
+                    if ":" in vsphere_vm_tag:
+                        tag_key_value = vsphere_vm_tag.split(":")
+                        tag_key = tag_key_value[0]
+                        tag_value = tag_key_value[1]
+                        if tag_key in filters and tag_value == filters[tag_key]:
+                            matched_tags[tag_key] = tag_value
+
+                        all_tags[tag_key] = tag_value
+
+                # Update the tag cache with latest tags
+                self.tag_cache[vm_id] = all_tags
+
+                if len(matched_tags) == len(filters):
```

Comment:
I think it's worth splitting out this logic into one or more helper functions and unit testing them

Reply:
Addressed